libcentrifugo/engineredis.go

```diff
@@ -200,12 +203,50 @@ end
 return n
 	`
 
+	// KEYS[1] - presence set key
+	// KEYS[2] - presence hash key
+	// ARGV[1] - key expire seconds
+	// ARGV[2] - expire at for set member
+	// ARGV[3] - uid
+	// ARGV[4] - info payload
+	addPresenceSource := `
+redis.call("zadd", KEYS[1], ARGV[2], ARGV[3])
+redis.call("hset", KEYS[2], ARGV[3], ARGV[4])
+redis.call("expire", KEYS[1], ARGV[1])
+redis.call("expire", KEYS[2], ARGV[1])
+	`
+
+	// KEYS[1] - presence set key
+	// KEYS[2] - presence hash key
+	// ARGV[1] - uid
+	remPresenceSource := `
+redis.call("hdel", KEYS[2], ARGV[1])
+redis.call("zrem", KEYS[1], ARGV[1])
+	`
+
+	// KEYS[1] - presence set key
+	// KEYS[2] - presence hash key
+	// ARGV[1] - now string
+	presenceSource := `
+local expired = redis.call("zrangebyscore", KEYS[1], "0", ARGV[1])
+if #expired > 0 then
+  for num = 1, #expired do
+    redis.call("hdel", KEYS[2], expired[num])
```

Comment:
Can't you use `zremrangebyscore` to do that in one command?

It has to do loop inside but much lower overhead for lua calls etc. Probably insignificant for common case when there will be only very short list of expired entries though